tgui/packages/tgui/interfaces/MusicInstrument.tsx

```diff
@@ -0,0 +1,183 @@
+      act('play_keyboard_on');
+      setKeybindToggle(true);
+    }
+  };
+
+  const keyIndexWithinRange = (index: number) => index + transpose >= 0 && index + transpose < notes.length;
+
+  const playNote = (index: number) => {
+    if (keyIndexWithinRange(index) && !activeKeys[index]) {
+      act('play_note', { note: index + transpose + 1, volume: volume });
+      const newKeys = activeKeys;
+      newKeys[index] = true;
+      setActiveKeys(newKeys);
+    }
+  };
+
+  const playNoteRelease = (index: number) => {
+    if (keyIndexWithinRange(index)) {
+      const newKeys = activeKeys;
+      newKeys[index] = false;
+      setActiveKeys(newKeys);
+    }
+  };
+
+  const getKeyboardIndex = (key: string) => {
+    return keyOffset + noteKeysOrder.findIndex((keyOrder) => keyOrder === key);
+  };
+
+  return (
+    <Window title={name} width={50 + notes.length * 30} height={410}>
+      <Window.Content
+        onKeyUp={(ev) => {
+          if (keybindToggle) {
+            let index = getKeyboardIndex(ev.key);
+            playNoteRelease(index);
+          }
+        }}
+        onKeyDown={(ev) => {
+          if (keybindToggle) {
+            let index = getKeyboardIndex(ev.key);
+            playNote(index);
+          }
+
+          if (ev.key === 'Control') {
+            toggleKeybind();
+          }
+        }}>
+        <Box className="instrument__keyboardwrapper">
+          <Box className="instrument__outerpanel">
+            <Box className="instrument__speaker" />
+            <Flex direction="column">
+              <Box className="instrument__panel">
+                <Box className="instrument__keyboardsupport">
+                  <Button
+                    className="instrument__toggle-keyboard-button"
+                    title="Toggle keyboard support (toggle with ctrl)"
+                    onClick={() => toggleKeybind()}
+                    icon="keyboard"
+                  />
+                  <Box
+                    className="instrument__keybind-indicator"
+                    style={{
+                      'box-shadow': `0px 0px 5px ${keybindToggle ? '#1b9b37' : '#5a1919'}`,
+                      'background': `${keybindToggle ? '#1b9b37' : '#5a1919'}`,
+                    }}
+                  />
+                </Box>
+                <Box className="instrument__panel-input">
+                  <Knob
+                    animated
+                    stepPixelSize={24}
+                    minValue={-24}
+                    maxValue={24}
+                    value={keyOffset}
+                    onDrag={(e, v) => setKeyOffset(v)}
+                    title={'Keybind offset'}
+                  />
+                  <span>Offset</span>
+                </Box>
+                <Box className="instrument_panel-info">
+                  <h1 style={{ 'text-align': 'center' }}>{name.toUpperCase()}</h1>
+                </Box>
+                <Box className="instrument__panel-input">
+                  <Knob
```

Comment:
`Knob` components should include `onChange` props in addition to `onDrag`, to support people editing via clicking and typing in the new value.

Reply:
onDrag actually accounts for that

Reply:
yep iirc